### DISCUSSION TITLE
Mac M1 Mach-o, but wrong architecture

### DISCUSSION THREAD

**Discussion by chaelbott** [explosion/spaCy#9561] [Help: Installation] [answered]
Hey all, I'm trying to install & use spacy on an M1 macbook air. I've installed what I need to with pip (to the best of my knowledge) but keep getting the same error when trying to run "$ python3 -m spacy download en_core_web_sm." Can anyone guide me in the right direction? I'll admit I'm not experienced with OSX at all, and may have a lot of questions/need step-by-step help.

Terminal reads:
```
michaelb@Michaels-MacBook-Air ~ % python3 -m spacy download en_core_web_sm
Traceback (most recent call last):
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/runpy.py", line 187, in _run_module_as_main
    mod_name, mod_spec, code = _get_module_details(mod_name, _Error)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/runpy.py", line 146, in _get_module_details
    return _get_module_details(pkg_main_name, error)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/runpy.py", line 110, in _get_module_details
    __import__(pkg_name)
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/spacy/__init__.py", line 11, in <module>
    from thinc.api import prefer_gpu, require_gpu, require_cpu  # noqa: F401
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/thinc/__init__.py", line 5, in <module>
    from .config import registry
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/thinc/config.py", line 14, in <module>
    import srsly
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/srsly/__init__.py", line 1, in <module>
    from ._json_api import read_json, read_gzip_json, write_json, write_gzip_json
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/srsly/_json_api.py", line 6, in <module>
    from . import ujson
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/srsly/ujson/__init__.py", line 1, in <module>
    from .ujson import decode, encode, dump, dumps, load, loads  # noqa: F401
ImportError: dlopen(/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/srsly/ujson/ujson.cpython-310-darwin.so, 2): no suitable image found.  Did find:
	/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/srsly/ujson/ujson.cpython-310-darwin.so: mach-o, but wrong architecture
	/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/srsly/ujson/ujson.cpython-310-darwin.so: mach-o, but wrong architecture
```

**Comment by polm** [comment]
It sounds like you have somehow mixed up the architecture of libraries you have installed, or something is weird with their Rosetta configuration. Maybe look at [this thread](https://github.com/apple/tensorflow_macos/issues/146#issuecomment-770349104) from January for a fix, or try uninstalling and resintalling `ujson` or all related packages.

**Reply by adrianeboyd** [reply]
Is this python from the universal2 installer?

At least for working with spacy, I would recommend NOT using a universal2 python installer and instead using an arm-only python binary to make it easier to maintain a venv without these kinds of mismatches. I've used python from homebrew or python from the miniforge osx arm64 installer with no issues.

You can use python from miniforge without using a conda environment, but if you do want to use conda, we provide osx arm64 packages for spacy on conda-forge, just run `conda install spacy -c conda-forge` in your new conda environment.

If you want to use this particular python, another (admittedly very annoying) option is to prefix every single python command with `arch -arm64`.

**Comment by adrianeboyd** [comment, marked as answer]
As an update, I think our pypi wheel releases were part of the problem here. There was a bug in our release process that led to macos x86_64 wheels for python 3.10 being named `universal2` instead of `x86_64`. `pip` on an M1 would happily install the `universal2` wheel but then it could only run correctly with `arch -x86_64 python` and would be incompatible with any installed `arm64` wheels.

We've replaced the incorrect `universal2` wheels with `x86_64` wheels for macos python 3.10 on pypi, so now when you install spacy and its dependencies it should build everything from source for `arm64` in a compatible way.

Edited to add: you may need delete the `universal2` wheels from your pip cache or add `--no-cache-dir` the next time you install the same version.
